### DISCUSSION TITLE
Calculating the Jacobian matrix of each layer separately

### DISCUSSION THREAD

**Discussion by nalzok** [google/flax#2265] [Q&A] [answered]
Neural networks are essentially nested functions. Take a 2-layer network for example, we can view each layer as a function, i.e.

$$
\begin{align}
f_1(X) =& \sigma(W_1 X + b_1) \\
f_2(X) =& W_2 X + b_2 \\
f(X) =& f_2(f_1(X)).
\end{align}
$$

While the neural network contains non-linear activations $\sigma$, we can "linearize" it at any particular point in the input space by considering the 1st order Taylor expansion.

$$
\begin{align}
f(X_0 + \delta) =& f_2(f_1(X_0 + \delta)) \\
=& f_2(f_1(X_0) + J_1 \delta)) \\
=& f_2(f_1(X_0))) + J_2 J_1 \delta,
\end{align}
$$

where $J_1$ is the Jacobian matrix of $f_1$ at $X_0$, and $J_2$ is the Jacobian matrix of $f_2$ at $f_1(X_0)$. Note that $J_1$ and $J_2$ are the Jacobian matrice of each layer with respect to the output activation of the previous layer.

I need to calculate $J_1$ and $J_2$ as part of a research project. I can easily find the product of $J_2$ and $J_1$ with something like `jax.grad(nn.apply, argnums=1)`, but how can I find $J_1$ and $J_2$ separately?

My idea is to use something like the following, but that requires me to manually split the neural network into separate layers. Is there a more automated way?

```python
class CNN(nn.Module):
    """A simple CNN model."""
    def setup(self):
        self.conv1 = nn.Conv(features=32, kernel_size=(3, 3))
        self.conv2 = nn.Conv(features=64, kernel_size=(3, 3))
        self.dense1 = nn.Dense(features=256)
        self.dense2 = nn.Dense(features=10)

    def __call__(self, x):
        return self.f4(self.f3(self.f2(self.f1(x))))

    def f1(self, x):
        x = self.conv1(x)
        return x

    def f2(self, x):
        x = nn.relu(x)
        x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
        x = self.conv2(x)
        return x

    def f3(self, x):
        x = nn.relu(x)
        x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
        x = x.reshape((x.shape[0], -1))  # flatten
        x = self.dense1(x)
        return x

    def f4(self, x):
        x = nn.relu(x)
        x = self.dense2(x)
        return x
```

**Comment by cgarciae** [comment, marked as answer]
Hey @nalzok, I think your scheme is the best practical solution for now.
One could dream of using interpreter magic to extract a sub-function from a complete function but I don't know if this has been done.
You could post this in the JAX repo and see if this is actually possible or not.
